main.go

```diff
@@ -623,20 +646,159 @@ func deleteSecret(kubeClient *k8s.Client, iamService *GoogleCloudIAMService, sec
+func watchServiceAccounts(waitGroup *sync.WaitGroup, kubeClientset *kubernetes.Clientset, iamService *GoogleCloudIAMService) {
+	// loop indefinitely
+	for {
+		log.Info().Msg("Watching serviceaccounts for all namespaces...")
+		timeoutSeconds := int64(300)
+
+		log.Info().Msg("Watching serviceaccount for all namespaces...")
+		watcher, err := kubeClientset.CoreV1().ServiceAccounts("").Watch(context.Background(), metav1.ListOptions{
+			TimeoutSeconds: &timeoutSeconds,
+		})
+
+		if err != nil {
+			log.Error().Err(err)
+		} else {
+			// loop indefinitely, unless it erros
+			for {
+				event, ok := <-watcher.ResultChan()
+				if !ok {
+					log.Warn().Msg("Watcher for secrets is closed")
+					break
+				}
+				if event.Type == watch.Added || event.Type == watch.Modified {
+					serviceAccount, ok := event.Object.(*v1.ServiceAccount)
+					if !ok {
+						log.Warn().Msg("Watcher for serviceaccount returns event object of incorrect type")
+						break
+					}
+					waitGroup.Add(1)
+					processServiceAccount(kubeClientset, iamService, serviceAccount, fmt.Sprintf("watcher:%v", event.Type))
+					waitGroup.Done()
+				}
+			}
+		}
+		// sleep random time between 22 and 37 seconds
+		sleepTime := foundation.ApplyJitter(30)
+		log.Info().Msgf("Sleeping for %v seconds...", sleepTime)
+		time.Sleep(time.Duration(sleepTime) * time.Second)
+	}
+}
+
+func getDesiredServiceAccountState(serviceAccount *v1.ServiceAccount) (state GCPServiceAccountState) {
+	var ok bool
+	// get annotations or set default value
+	state.Enabled, ok = serviceAccount.Annotations[annotationGCPServiceAccount]
+	if !ok {
+		state.Enabled = "false"
+	}
+
+	state.Name, ok = serviceAccount.Annotations[annotationGCPServiceAccountName]
+	if !ok {
+		state.Name = ""
+	}
+
+	return
+}
+
+func getCurrentServiceAccountState(serviceAccount *v1.ServiceAccount) (state GCPServiceAccountState) {
+
+	// get state stored in annotations if present or set to empty struct
+	gcpServiceAccountStateString, ok := serviceAccount.Annotations[annotationGCPServiceAccountState]
+	if !ok {
+		// couldn't find saved state, setting to default struct
+		state = GCPServiceAccountState{}
+	}
+
+	if err := json.Unmarshal([]byte(gcpServiceAccountStateString), &state); err != nil {
+		// couldn't deserialize, setting to default struct
+		state = GCPServiceAccountState{}
+		return
+	}
+
+	// return deserialized state
+	return
+}
+
+func makeServiceAccountChanges(kubeClientset *kubernetes.Clientset, iamService *GoogleCloudIAMService, serviceAccount *v1.ServiceAccount, initiator string, desiredState, currentState GCPServiceAccountState) (err error) {
+
+	// parse last attempt time from state
+	lastAttempt := time.Time{}
+	if currentState.LastAttempt != "" {
+		var err error
+		lastAttempt, err = time.Parse(time.RFC3339, currentState.LastAttempt)
+		if err != nil {
+			lastAttempt = time.Time{}
+		}
+	}
+	if desiredState.Enabled == "true" && desiredState.Name != "" && time.Since(lastAttempt).Minutes() > 15 {
```

Comment:
You want to avoid this from getting executed every time the service account changes, so you probably want to add the following to the condition:

```go
&& currentState.FullServiceAccountEmail == ""
```